packages/flutter_tools/lib/src/application_package.dart

```diff
@@ -120,11 +120,25 @@ class AndroidApk extends ApplicationPackage {
     final String packageId = manifests.first.getAttribute('package');
 
     String launchActivity;
-    for (xml.XmlElement category in document.findAllElements('category')) {
-      if (category.getAttribute('android:name') == 'android.intent.category.LAUNCHER') {
-        final xml.XmlElement activity = category.parent.parent;
-        final String enabled = activity.getAttribute('android:enabled');
-        if (enabled == null || enabled == 'true') {
+    for (xml.XmlElement activity in document.findAllElements('activity')) {
+      final String enabled = activity.getAttribute('android:enabled');
+      if(enabled!=null && enabled!='true')
+        continue;
+      for(xml.XmlElement element in activity.findElements('intent-filter')){
+        String actionName = '', categoryName = '';
+        for(xml.XmlNode node in element.children){
+          if(!(node is xml.XmlElement))
+            continue;
+          final xml.XmlElement xmlElement = node;
+          final String name = xmlElement.getAttribute('android:name');
+          if(name == 'android.intent.action.MAIN'){
+            actionName = name;
+          }
+          else if(name == 'android.intent.category.LAUNCHER'){
+            categoryName = name;
+          }
+        }
+        if(actionName.isNotEmpty && categoryName.isNotEmpty){
```

Comment:
How can this conditional be true?  Up above you use an `if, else if` to assign these variables.  So only one of them will ever be non-empty, right?

Reply:
Also, please add appropriate spaces: `if (....) {`